include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h

```diff
@@ -275,7 +219,18 @@ struct is_hetero_legacy_trait<Iter, ::std::enable_if_t<Iter::is_hetero::value>>
 };
 
 template <typename Iter>
-inline constexpr bool is_sycl_iterator_v = is_sycl_iterator<Iter>::value || is_hetero_legacy_trait<Iter>::value;
+using is_sycl_or_hetero_iterator = std::disjunction<is_sycl_iterator<Iter>, is_hetero_legacy_trait<Iter>>;
+
+template <typename Iter>
+inline constexpr bool is_sycl_or_hetero_iterator_v = is_sycl_or_hetero_iterator<Iter>::value;
+
+template <typename _Iter>
+using __is_passed_directly_device_ready =
+    std::conjunction<oneapi::dpl::is_indirectly_device_accessible<_Iter>, sycl::is_device_copyable<_Iter>,
+                     oneapi::dpl::__internal::__is_random_access_iterator<_Iter>>;
```

Comment:
Do we want to require that the iterator is random access in the spec? Here's the corresponding comment: https://github.com/uxlfoundation/oneAPI-spec/pull/620#discussion_r2069304471

Probably, we should choose to change it in the spec PR or remove it here.

Reply:
I think its something we want.  However, I'm not so confident that checking for `std::random_access_iterator_tag` is a super reliable way of doing this, but I think it is the best we have `pre-c++20` concepts.

I worry a little about this cutting off custom types which are in practice random access iterators, but do not add the random access iterator tag.  

@akukanov, Should we enforce this requirement, or only state it in the specification without enforcement?

Reply:
It's really orthogonal to the iterators, device accessibility etc. - it's a question of the requirements set for/by parallel algorithms, specifically when invoked with a device policy. It probably deserves clarification, but not within the scope of this patch nor the corresponding spec patch.

Reply:
I suppose so.  We are already covered for `sycl_iterator` by a lack of device copyability of a `sycl::queue`, so maybe this check is frivolous and just liable to get us into trouble.  I think I will remove it.